src/cpl/share_esmf/lnd_set_decomp_and_domain.F90

```diff
@@ -0,0 +1,881 @@
+       mincornerCoord(2) = yv(1,1,1)   ! min lat
+       maxcornerCoord(1) = xv(3,ni,nj) ! max lon
+       maxcornerCoord(2) = yv(3,ni,nj) ! max lat
+       deallocate(xv,yv)
+
+    end if
+
+    ! close file
+    call pio_seterrorhandling(pioid, PIO_INTERNAL_ERROR)
+    call pio_closefile(pioid)
+
+    ! create the ESMF grid
+    lgrid = ESMF_GridCreateNoPeriDimUfrm (maxindex=maxindex, &
+         mincornercoord=mincornercoord, maxcornercoord= maxcornercoord, &
+         staggerloclist=(/ESMF_STAGGERLOC_CENTER, ESMF_STAGGERLOC_CORNER/), rc=rc)
+    if (ChkErr(rc,__LINE__,u_FILE_u)) return
+
+    ! create the mesh from the lgrid
+    mesh_lndcreate =  ESMF_MeshCreate(lgrid, rc=rc)
+    if (ChkErr(rc,__LINE__,u_FILE_u)) return
+
+    ! Set the mesh mask to 1
+    call ESMF_MeshGet(mesh_lndcreate, spatialDim=spatialDim, numOwnedElements=numOwnedElements, rc=rc)
+    if (chkerr(rc,__LINE__,u_FILE_u)) return
+    allocate(ownedElemCoords(spatialDim*numownedelements))
+    call ESMF_MeshGet(mesh_lndcreate, ownedElemCoords=ownedElemCoords, rc=rc)
+    if (ChkErr(rc,__LINE__,u_FILE_u)) return
+    allocate(lnd_mask(numownedelements))
+    lnd_mask(:) = 1
+    ! call ESMF_MeshSet(mesh_lndcreate, elementMask=lnd_mask, rc=rc)
+    ! if (ChkErr(rc,__LINE__,u_FILE_u)) return
+
+    ! Determine ldecomp and ldomain
+    call decompInit_lnd(lni=ni, lnj=nj, amask=lnd_mask)
+    if (use_soil_moisture_streams) then
+       call decompInit_lnd3D(lni=ni, lnj=nj, lnk=nlevsoi)
+    end if
+
+    ! Initialize processor bounds
+    call get_proc_bounds(bounds)
+    begg = bounds%begg
+    endg = bounds%endg
+
+    ! Create gindex_ctsm
+    nlnd = endg - begg + 1
+    allocate(gindex_ctsm(nlnd))
+    do g = begg, endg
+       n = 1 + (g - begg)
+       gindex_ctsm(n) = ldecomp%gdc2glo(g)
+    end do
+
+    ! Initialize domain data structure
+    isgrid2d = .true.
+    call domain_init(domain=ldomain, isgrid2d=isgrid2d, ni=ni, nj=nj, nbeg=begg, nend=endg)
+
+    ! Determine ldomain%mask and ldomain%frac
+    do g = begg, endg
+       ldomain%mask(g) = 1
+       ldomain%frac(g) = 1._r8
+    end do
+
+    ! Generate a new mesh on the gindex decomposition
+    distGrid_ctsm = ESMF_DistGridCreate(arbSeqIndexList=gindex_ctsm, rc=rc)
+    if (ChkErr(rc,__LINE__,u_FILE_u)) return
+    mesh_ctsm = ESMF_MeshCreate(mesh_lndcreate, elementDistGrid=distgrid_ctsm, rc=rc)
+    if (ChkErr(rc,__LINE__,u_FILE_u)) return
+
+    ! Set ldomain%lonc, ldomain%latc and ldomain%area
+    call lnd_set_ldomain_gridinfo_from_mesh(mesh_ctsm, vm, gindex_ctsm, begg, endg, isgrid2d, ni, nj, ldomain, rc)
+    if (ChkErr(rc,__LINE__,u_FILE_u)) return
+
+    deallocate(lnd_mask)
+
+  end subroutine lnd_set_decomp_and_domain_from_createmesh
+
+  !===============================================================================
+  subroutine lnd_get_global_dims(ni, nj, gsize, isgrid2d)
+
+    ! Determine global 2d sizes from read of dimensions of surface dataset
```

Comment:
- [ ] This subroutine is probably the least intuitive part of these changes. You gave me a good verbal explanation for why it's needed, but I think it will really help us when we look back at this in a few years if that explanation is recorded in the comment here: both (a) why we need the 2-d sizes, and (b) why we need to resort to the surface dataset to get this information.

Reply:
Can you please add this comment or something like it?

`Meshes do not indicate if the mesh can be represented as a logically rectangular grid. However, CTSM needs this information in the history file generation via the logical variable isgrid2d. Since for CMEPS and LILAC there is no longer the need for the fatmlndfrc file (where the isgrid2d variable was determined from before) - the surface dataset is now used to determine if the underlying grid is 2d or not.`